activesupport/lib/active_support/cache/coder.rb

```diff
@@ -0,0 +1,142 @@
+      end
+
+      def dump_compressed(entry, threshold)
+        return @serializer.dump_compressed(entry, threshold) if @legacy_serializer
+
+        # If value is a string with a supported encoding, use it as the payload
+        # instead of passing it through the serializer.
+        if type = type_for_string(entry.value)
+          payload = binary_compatible_string(entry.value)
+        else
+          type = OBJECT_DUMP_TYPE
+          payload = serialize(entry.value)
+        end
+
+        # Compress payload if it meets the size threshold. (Note that `type < 0`
+        # signifies a compressed payload with `type.abs` as its actual type.)
+        if compressed = try_compress(payload, threshold)
+          payload = compressed
+          type = -type
+        end
+
+        expires_at = entry.expires_at || -1.0
+
+        version = dump_version(entry.version)
+        version_length = version&.bytesize || -1
+
+        packed = [SIGNATURE, type, expires_at, version_length].pack(PACKED_TEMPLATE)
+        packed << version if version
+        packed << payload
+      end
+
+      def load(dumped)
+        return @serializer.load(dumped) if !signature?(dumped)
+
+        type = dumped.unpack1(PACKED_TYPE_TEMPLATE)
+        expires_at = dumped.unpack1(PACKED_EXPIRES_AT_TEMPLATE)
+        version_length = dumped.unpack1(PACKED_VERSION_LENGTH_TEMPLATE)
+
+        expires_at = nil if expires_at < 0
+        version = load_version(dumped.byteslice(PACKED_VERSION_INDEX, version_length))
+        payload = dumped.byteslice((PACKED_VERSION_INDEX + [version_length, 0].max)..)
+
+        payload = decompress(payload) if type < 0
+
+        if string_encoding = STRING_ENCODINGS[type.abs]
+          value = payload.force_encoding(string_encoding)
+        else
+          value = deserialize(payload)
+        end
+
+        Cache::Entry.new(value, version: version, expires_at: expires_at)
+      end
+
+      private
+        SIGNATURE = "\x00\x11".b.freeze
+
+        OBJECT_DUMP_TYPE = 0x01
+
+        STRING_ENCODINGS = {
+          0x02 => Encoding::UTF_8,
+          0x03 => Encoding::BINARY,
+          0x04 => Encoding::US_ASCII,
+        }
+
+        PACKED_TEMPLATE = "A#{SIGNATURE.bytesize}cEl<"
+        PACKED_TYPE_TEMPLATE = "@#{SIGNATURE.bytesize}c"
+        PACKED_EXPIRES_AT_TEMPLATE = "@#{[0].pack(PACKED_TYPE_TEMPLATE).bytesize}E"
+        PACKED_VERSION_LENGTH_TEMPLATE = "@#{[0].pack(PACKED_EXPIRES_AT_TEMPLATE).bytesize}l<"
+        PACKED_VERSION_INDEX = [0].pack(PACKED_VERSION_LENGTH_TEMPLATE).bytesize
+
+        MARSHAL_SIGNATURE = "\x04\x08".b.freeze
+
+        def signature?(dumped)
+          dumped.is_a?(String) && dumped.start_with?(SIGNATURE)
+        end
+
+        def type_for_string(value)
+          STRING_ENCODINGS.key(value.encoding) if value.instance_of?(String)
+        end
+
+        def binary_compatible_string(string)
+          (string.encoding == Encoding::BINARY || string.ascii_only?) ? string : string.b
```

Comment:
I'm not sure this saves much. Any reason why you don't just call `String#b` instead of this helper?

Or perhaps `s.encoding == Encoding::BINARY ? s : s.b` if we want to save an allocation, but binary strings would be rare in the first place. Also `b` creates a shared string, so it's only allocating an object slot, not copying the bytes.

`#ascii_only?` triggers a coderange scan, so it may be detrimental if the string is long.

Reply:
I mostly agree.  I added this due to https://github.com/rails/rails/pull/48751#discussion_r1266247836.  The benchmark in that thread favors just calling `String#b`, but I started to wonder if there might be some compounding effect of GC pressure in a high load environment that the benchmark doesn't capture.

Anyway, that's just idle speculation, so I've made this change.

Reply:
> compounding effect of GC pressure in a high load environment 

I think we're talking of something like 2 strings per call to `.b` (if the original string is mutable, if not, only one).

I don't think that's a big problem for the GC. But yeah, I wish Ruby had a better API to do buffers like this, I might open a ticket upstream.